### PR TITLE
Retry logic for S3 bucket client

### PR DESCRIPTION
**Pull request by alexqyle**
**What this PR does**:
* Created S3 bucket client wrapper to have retry logic for any bucket client failures. Retry could be configured.

**Which issue(s) this PR fixes**:
Fixes #<issue number>

**Checklist**
- [x] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: docs/blocks-storage/querier.md [hunk 1]
```diff
@@ -277,6 +277,18 @@ blocks_storage:
...
+    [operation_retries: <int> | default = 1]
+
+    # The minimum backoff time when retrying failed AWS S3 operation.
+    # CLI flag: -blocks-storage.s3.retry-min-backoff
+    [retry_min_backoff: <duration> | default = 10s]
+
+    # The maximum backoff time when retrying failed AWS S3 operation.
+    # CLI flag: -blocks-storage.s3.retry-max-backoff
+    [retry_max_backoff: <duration> | default = 1m]
+
```
Comment (alanprot) [inline docs/blocks-storage/querier.md:291]: Does this needs to be configurable? 

We are trying to have fewer configs so if we could avoid adding configs that probably nobody will change i think its better...

@yeya24 WDYT?

Comment (yeya24) [reply docs/blocks-storage/querier.md:291]: Agree. Let's find a sane default value and expose these configs in the future if needed.

File: pkg/storage/bucket/s3/bucket_client.go [hunk 2]
```diff
@@ -1,6 +1,12 @@
 package s3
 
 import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util/backoff"
```
Comment (yeya24) [inline pkg/storage/bucket/s3/bucket_client.go:8]: Please group the import. Cortex package should be at the last section separated from other packages we import.

File: pkg/compactor/shuffle_sharding_planner.go [hunk 3]
```diff
@@ -79,7 +80,8 @@ func (p *ShuffleShardingPlanner) Plan(_ context.Context, metasByMinTime []*metad
 			return nil, fmt.Errorf("unable to get visit marker file for block %s: %s", blockID, err.Error())
 		}
 		if !blockVisitMarker.isVisitedByCompactor(p.blockVisitMarkerTimeout, p.ringLifecyclerID) {
-			return nil, fmt.Errorf("block %s is not visited by current compactor %s", blockID, p.ringLifecyclerID)
+			level.Warn(p.logger).Log("msg", "block is not visited by current compactor", "block_id", blockID, "compactor_id", p.ringLifecyclerID)
+			return nil, nil
```
Comment (yeya24) [inline pkg/compactor/shuffle_sharding_planner.go:84]: Would you like to elaborate more on this change? Seems unrelated to the retry logic change here.

Comment (alexqyle) [reply pkg/compactor/shuffle_sharding_planner.go:84]: I will create a separate PR for this

**Review by yeya24 — APPROVED** [review]
Overall this change looks good to me. Thanks!
Probably 2 min is a little bit too much for max retry as usually a single query timeout is < 2m.
